ioredis_streaming_scan.js

```diff
@@ -23,42 +23,95 @@ const createLargeSet = async () => {
 };
 
 const ioRedisRegularSetScan = async () => {
-  let it = '0';
+  let cursor = '0'; // Store cursor value from Redis.
   const allMembers = [];
 
   do {
-    const response = await redis.sscan(LARGE_SET_KEY, it);
-    // Get next value of cursor and log responses.
-    it = response[0];
-    const setMembers = response[1];
-    allMembers.push(...setMembers);
-    console.log(setMembers);
-  } while (it !== '0');
-
-  console.log(`Done scanning the large set, read ${allMembers.length} members.`);
+    const response = await redis.sscan(LARGE_SET_KEY, cursor);
+    // Get next value of cursor and store values returned.
+    cursor = response[0];
+    allMembers.push(...response[1]);
+  } while (cursor !== '0');  // Done when cursor is '0' again.
+
+  return allMembers;
 };
 
 const ioRedisStreamingSetScan = async () => {
-  // Read back all set members, using a scan approach (SSCAN).
-  const allMembers = [];
-  const stream = redis.sscanStream(LARGE_SET_KEY);
+  return new Promise ((resolve, reject) => {
+    try {
+      // Read back all set members, using a scan approach (SSCAN).
+      const allMembers = [];
+      const stream = redis.sscanStream(LARGE_SET_KEY);
 
-  stream.on('data', function (setMembers) {
-    // setMembers is an array of 1 or more members.
-    allMembers.push(...setMembers);
-    console.log(setMembers);
-  });
+      stream.on('data', function (setMembers) {
+        // setMembers is an array of 1 or more members.
+        allMembers.push(...setMembers);
+      });
 
-  stream.on('end', function () {
-    console.log(`Done scanning the large set, read ${allMembers.length} members.`);
-    redis.quit();
+      stream.on('end', function () {
+        // No more members left to read.
+        resolve(allMembers);
+      });
+    } catch (e) {
+      // Something went wrong :(
+      reject(e);
+    }
   });
 };
 
+const ioRedisES6GeneratorSetScan = async () => {
+  // Read back all set members, using ES6 Generator function approach.
+  const setMembersGenerator = async function* () {
+    let cursor = '0';
+
+    do {
+      const response = await redis.sscan(LARGE_SET_KEY, cursor);
+
+      // Get next value of cursor.
+      cursor = response[0];
+
+      // Yield the set members returned from Redis.
+      yield response[1];
+    } while (cursor !== '0');
+  };
+
+  // Get values from the generator.  Using the generator shields us 
+  // from the details of the Redis implementation.
+
+  const allMembers = [];
+  const setIterator = setMembersGenerator();
+  let done = false;
+
+  do {
+    let setMembers;
+
+    ({ value: setMembers, done } = await setIterator.next());
```

Comment:
While these parens are syntactically correct, they are also not required. Personally, and I know this is an opinion, I think they look weird.

Reply:
Turned out we needed them as there's a syntax error without - `{ value: setMembers, done}` is interpreted as an object declaration not an assignment.  I was surprised by this!

Reply:
Simplified this down to its bones as per Slack conversation.